diff --git a/java/com/craftinginterpreters/lox/Interpreter.java b/java/com/craftinginterpreters/lox/Interpreter.java
index 823503f4..8c638059 100644
--- a/java/com/craftinginterpreters/lox/Interpreter.java
+++ b/java/com/craftinginterpreters/lox/Interpreter.java
@@ -2,6 +2,10 @@
 package com.craftinginterpreters.lox;
 //> Statements and State import-list
 
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+
 //> Functions import-array-list
 import java.util.ArrayList;
 //< Functions import-array-list
@@ -35,6 +39,9 @@ class Interpreter implements Expr.Visitor<Object>,
 
 //< Statements and State environment-field
 //> Functions interpreter-constructor
+
+  private InputStreamReader getcStream = new InputStreamReader(System.in, StandardCharsets.UTF_8);
+
   Interpreter() {
     globals.define("clock", new LoxCallable() {
       @Override
@@ -49,8 +56,74 @@ class Interpreter implements Expr.Visitor<Object>,
       @Override
       public String toString() { return "<native fn>"; }
     });
+
+    globals.define("getc", new LoxCallable() {
+      @Override
+      public int arity() { return 0; }
+
+      @Override
+      public Object call(Interpreter interpreter,
+                         List<Object> arguments) {
+        try {
+          int c = getcStream.read();
+          if (c < 0) {
+            return (double)-1;
+          }
+          return (double)c;
+        } catch (IOException error) {
+          return (double)-1;
+        }
+      }
+
+      @Override
+      public String toString() { return "<native fn>"; }
+    });
+
+    globals.define("chr", new LoxCallable() {
+      @Override
+      public int arity() { return 1; }
+
+      @Override
+      public Object call(Interpreter interpreter,
+                         List<Object> arguments) {
+        return Character.toString((char)(double)arguments.get(0));
+      }
+
+      @Override
+      public String toString() { return "<native fn>"; }
+    });
+
+    globals.define("exit", new LoxCallable() {
+      @Override
+      public int arity() { return 1; }
+
+      @Override
+      public Object call(Interpreter interpreter,
+                         List<Object> arguments) {
+        System.exit((int)(double)arguments.get(0));
+        return null;
+      }
+
+      @Override
+      public String toString() { return "<native fn>"; }
+    });
+
+    globals.define("print_error", new LoxCallable() {
+      @Override
+      public int arity() { return 1; }
+
+      @Override
+      public Object call(Interpreter interpreter,
+                         List<Object> arguments) {
+        System.err.println((String)arguments.get(0));
+        return null;
+      }
+
+      @Override
+      public String toString() { return "<native fn>"; }
+    });
   }
-  
+
 //< Functions interpreter-constructor
 /* Evaluating Expressions interpret < Statements and State interpret
   void interpret(Expr expression) { // [void]
